### PR TITLE
Profile and settings  modal

### LINKED ISSUE #19: My profile modal
Should allow for updating a subset of fields from the `User` schema.

Should allow user to update their profile picture by uploading to a blob store in Vercel (hmu to set up store).

### PR DESCRIPTION
**Pull request by AlexanderWangY**
## Description
This PR adds a settings modal to the Competition Dashboard and the Admin Dashboard. This allows for users to access their account settings as well as leaving space for extra additions to the settings modal without hassle.

A User profile and account tab was implemented which allows for Avatar changes through _Vercel Blob_. It also includes the ability to edit the profile and account information (but not email) as well as add skills and a bio.

The Prisma scheme was merged and reflects the new User profile additions such as bio, skills, github and linkedin url.

A yup validator is used for the form used in the user profile settings form.

## Issues
Fixes #19 

## Type of Change
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Documentation update

## Additional Notes:
I do NOT like the new admin dashboard. :)

### CODE REVIEW

**Comment by AlexanderWangY** [comment]
@RobertConde Whats the timeline on this PR?

**Comment by RobertConde** [comment]
24 hrs

File: components/settings/tabs/Account.tsx [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+  const [formChanged, setFormChanged] = useState<boolean | null>(null);
+
+  const displayPhone = (phone: string) => {
+    if (!phone || phone.length !== 10) return null;
+
+    return `(${phone.slice(0, 3)})-${phone.slice(3, 6)}-${phone.slice(6)}`;
+  };
+
+  // Form initilization
+  const form = useForm({
```
Comment (RobertConde) [inline components/settings/tabs/Account.tsx:80]: You should do something like `useForm<T>` to specify the type of the form. This also is automatic by specifying initial values I believe (check Configure Competition).

File: components/settings/tabs/Account.tsx [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+    validate: yupResolver(profileConfigurationScheme),
+    onValuesChange: (values: any) => {
+      console.log(values);
+      if (values.bio !== null) setBioLength(values.bio.length);
+      setFormChanged(true);
+    },
+  });
+
+  // Preprocessing form values after submit
+  const cleanFormValues = (values: FormValues): Partial<FormValues> => {
```
Comment (RobertConde) [inline components/settings/tabs/Account.tsx:95]: You can add `transformValues: (vals) => ...` in the `useForm` hook instead. TS is not happy even though ik what you're doing.

File: components/settings/tabs/Account.tsx [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+    if (newUserData) {
+      // Updating session to reflect new user
+      await update({
+        ...session,
+        user: newUserData,
+      });
+
+      // Form resetting and state resetting
+      form.reset();
+      form.setFieldValue('bio', newUserData.bio || '');
```
Comment (RobertConde) [inline components/settings/tabs/Account.tsx:145]: Maybe a function should do a bunch of this shared behavior

File: components/settings/tabs/Account.tsx [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+      });
+    }
+
+    close();
+  };
+
+  const onUploadAvatar = async (event: any) => {
+    event.preventDefault();
+
+    if (session?.user?.id === undefined) {
```
Comment (RobertConde) [inline components/settings/tabs/Account.tsx:179]: nit: `!session?.user?.id`

File: components/settings/tabs/Account.tsx [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+        icon: <IconX />,
+        color: 'red',
+        autoClose: 3000,
+      });
+      return;
+    }
+
+    setLoadingAvatar(true);
+
+    await deleteUserAvatar(session.user.id);
```
Comment (RobertConde) [inline components/settings/tabs/Account.tsx:219]: shouldn't need to be called.

File: actions/user.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+const updateUserAvatar = async (user_id: string, formData: FormData) => {
+  const avatar_image = formData.get('file') as File;
+
+  const blob = await put(avatar_image.name, avatar_image, {
+    access: 'public',
+  });
+
+  revalidatePath('/');
+
```
Comment (RobertConde) [inline actions/user.ts:31]: You should delete previous avatar before before uploading and setting new one here! That way we don't forget to call "`delete`".

File: components/admin/AdminDashboard.tsx [hunk 7]
```diff
@@ -19,8 +28,11 @@ interface Props {
 
 export default function AdminDashboard({ links, competitions }: Props) {
   const combobox = useCombobox();
+  const { data: session, status } = useSession();
   const router = useRouter();
 
+  if (status === 'loading') return <Spinner />;
```
Comment (RobertConde) [inline components/admin/AdminDashboard.tsx:34]: You don't need this... nor dou need the `useSession` hook.

File: components/settings/tabs/Account.tsx [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+    const updatedUserData = await updateUserAvatar(
+      session?.user?.id!,
+      formData
+    );
```
Comment (RobertConde) [inline components/settings/tabs/Account.tsx:223]: Should be `formData.get('file') as File`

File: actions/user.ts [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    return new_user;
+  } catch {
+    return null;
+  }
+};
+
+const updateUserAvatar = async (user_id: string, formData: FormData) => {
+  const avatar_image = formData.get('file') as File;
+
+  const blob = await put(avatar_image.name, avatar_image, {
```
Comment (RobertConde) [inline actions/user.ts:26]: Should pass a `File`... not the form

File: actions/user.ts [hunk 10]
```diff
@@ -0,0 +1,67 @@
+'use server';
+
+import prisma from '@/prisma';
+import { type User } from '@prisma/client';
+import { del, put } from '@vercel/blob';
+import { revalidatePath } from 'next/cache';
+
```
Comment (RobertConde) [inline actions/user.ts:7]: Define return types to make sure you do what you expect!

File: actions/user.ts [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+const updateUserProfile = async (user_id: string, user: User) => {
+  try {
+    const new_user = await prisma.user.update({
+      where: {
```
Comment (RobertConde) [inline actions/user.ts:11]: You don't have to await this. You can just return the Promise itself

File: actions/user.ts [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+  const blob = await put(avatar_image.name, avatar_image, {
+    access: 'public',
+  });
+
+  revalidatePath('/');
+
+  try {
+    if (!blob.url) throw new Error('Missing url');
+    const new_user = await prisma.user.update({
+      where: {
```
Comment (RobertConde) [inline actions/user.ts:35]: Same just return the promise

File: actions/user.ts [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+  if (!user?.image) {
+    return;
+  }
+
+  try {
+    await del(user.image);
+  } catch {
+    console.log('Could not delete');
+  }
```
Comment (RobertConde) [inline actions/user.ts:64]: Is probably more descriptive `console.error('Could not delete avatar for user with ID ${user_id}');`

File: components/settings/tabs/Account.tsx [hunk 14]
```diff
@@ -0,0 +1,442 @@
...
+    return Object.fromEntries(cleanedEntries.filter((entry) => entry !== null));
+  };
+
+  // Submitting Form
+  const onSubmitForm = async (values: any) => {
+    open();
+
+    if (!session?.user?.id) {
+      console.error(
+        '[ERROR] No session ID. Please contact Technical Staff for help'
```
Comment (RobertConde) [inline components/settings/tabs/Account.tsx:126]: People shouldn't be reading this, generally. It could just be "Cannot submit form without a session ID"

**Comment by RobertConde** [comment]
Don't need migrations, yet. Remove then.

File: components/settings/SettingsModal.tsx [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+
+  return (
+    <Modal
+      opened={opened}
+      onClose={() => {
+        closeModal();
+        setActiveTab('account');
+      }}
+      size='lg'
+      fullScreen={isMobile ? true : false}
```
Comment (RobertConde) [inline components/settings/SettingsModal.tsx:103]: if true, return true .... hmmm

File: components/settings/SettingsModal.tsx [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+interface TabPanelItem {
+  value: string;
+  component: React.ComponentType;
+}
+
+export default function SettingsModal({ opened, closeModal }: SettingsProps) {
+  const iconStyle = { width: rem(20), height: rem(20) };
+  const isMobile = useMediaQuery(`(max-width: 50em)`);
+  const [activeTab, setActiveTab] = useState('account');
+  const [burgerOpen, { toggle, open: openBurger, close: closeBurger }] =
```
Comment (RobertConde) [inline components/settings/SettingsModal.tsx:46]: you're not using `openBurger`

**Comment by RobertConde** [comment]
Make sure to run `npm run format`

**Review by RobertConde — CHANGES_REQUESTED** [review]
okie

File: components/settings/SettingsModal.tsx (hunk 15 above)
Comment (AlexanderWangY) [reply components/settings/SettingsModal.tsx:103]: HAHAHA 😭

File: actions/user.ts [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+};
+
+const updateUserAvatar = async (user_id: string, formData: FormData) => {
+  const avatar_image = formData.get('file') as File;
+
+  const blob = await put(avatar_image.name, avatar_image, {
+    access: 'public',
+  });
+
+  revalidatePath('/');
```
[the remark this answers is not in the document]
Comment (AlexanderWangY) [reply actions/user.ts:30]: I just need to re-validate it for everyone (especially if this modal is going to be for both admins and hackers). Therefore I've decided to go with a whole application re-validation.

**Comment by RobertConde** [comment]
You did it!

**Pull request closed without merging**